### PR TITLE
[Feature][Abstract]: Settings & Config

### PR DESCRIPTION
**Pull request by joowon-dm-snu**
### 🚀  Related Issue

Abstract PR for solving #15 

### :tada: Details

they are basic classes for manta-client

### ☑️ Check lists

 - [x] config
 - [x] settings
 - [x] errors
 - [x] utils
 
#### Before Implementation
 - [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
 - [x] Did you write abstract codes for the new feature?
 - [x] Did you add documentations for upcoming feature?
 - [x] Did you write necessary tests?
 - [x] Did you make sure your PR does only one thing?
 - [x] Ready for implemtation?. Please ask for a review!

#### After Implementation
 - [ ] Did you verify new and existing tests pass locally with your changes?
 - [ ] Did you make sure to update the documentation with your changes? (if necessary)
 - [ ] Did you update the CHANGELOG? (not for typos, docs, test updates, or internal minor changes/refactorings)
 - [ ] Ready for merge?

### CODE REVIEW

File: manta_client/__init__.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import manta_client as mc
```
Comment (BlueWhaleKo) [inline manta_client/__init__.py:1]: It'd be nice if we could shorten our sdk name to `manta`

File: manta_client/base/config.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    def _load_defaults(self):
+        conf_dict = util.read_config_yaml("config_defaults.yaml")
+        if conf_dict is not None:
+            self.update(conf_dict)
+
+    def _assert_dict_values(self, v: Any) -> None:
+        raise NotImplementedError()
+
+    def _sanitize(self, k: str, v: Any) -> Tuple:
+        k = k.rstrip("_|-")
```
Comment (BlueWhaleKo) [inline manta_client/base/config.py:76]: need some documentation here :)

File: manta_client/base/config.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+        self._load_defaults()
+
+    def _load_defaults(self):
+        conf_dict = util.read_config_yaml("config_defaults.yaml")
+        if conf_dict is not None:
+            self.update(conf_dict)
+
+    def _assert_dict_values(self, v: Any) -> None:
+        raise NotImplementedError()
```
Comment (BlueWhaleKo) [inline manta_client/base/config.py:73]: need some documentation here :)

File: manta_client/base/config.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        return [k for k in self._items.keys() if not k.startswith("_")]
+
+    def values(self):
+        return [v for k, v in self._items.items() if not k.startswith("_")]
+
+    def items(self):
+        return [(k, v) for k, v in self._items.items() if not k.startswith("_")]
+
+    def set_callback(self, fn):
+        self._callback = fn
```
Comment (BlueWhaleKo) [inline manta_client/base/config.py:132]: need function interface for callback function.

I see they are used as  `self._callback(data=data)`, `self._callback(key=k, val=v)`.
It'd be nice if you could add some descriptions about the callback function

File: manta_client/base/setting.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+    def _public_keys(self):
+        raise NotImplementedError()
+
+    def _private_keys(self):
+        raise NotImplementedError()
+
+    @classmethod
+    def _property_keys(cls) -> Generator[str, None, None]:
+        return (k for k, v in vars(cls).items() if isinstance(v, property))
```
Comment (BlueWhaleKo) [inline manta_client/base/setting.py:76]: property seems not defined : (

File: manta_client/base/setting.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+    @classmethod
+    def _class_keys(cls) -> Generator[str, None, None]:
+        return (
+            k
+            for k, v in vars(cls).items()
+            if not k.startswith("_") and not callable(v) and not isinstance(v, property)
+        )
```
Comment (BlueWhaleKo) [inline manta_client/base/setting.py:84]: property seems not defined : (

File: manta_client/base/config.py (hunk 2 above)
Comment (BlueWhaleKo) [inline manta_client/base/config.py:76]: need some documentation here :)

File: manta_client/base/config.py (hunk 3 above)
Comment (BlueWhaleKo) [inline manta_client/base/config.py:73]: need some documentation here :)

File: manta_client/base/config.py (hunk 4 above)
Comment (BlueWhaleKo) [inline manta_client/base/config.py:132]: need function interface for callback

I see use cases as mentions below. But, I'm not sure how the callback function should look like

``` python
self._callback(key=k, val=v)
self._callback(data=data)
```

File: manta_client/util.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+                result.update(param)
+    except OSError:
+        raise Error("Couldn't read yaml file: %s" % path)
+    except UnicodeDecodeError:
+        print("wrong encoding")
+    return result
+
+
+def read_config_yaml(path: Union[str, Path], keyname: str = "value") -> Dict:
+    # TODO: (kjw) this process can be efficient but think this is not a time-consuming issue
```
Comment (BlueWhaleKo) [inline manta_client/util.py:33]: agreed

File: manta_client/util.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        result[k] = v[keyname]
+    return result
+
+
+def save_yaml(path: Union[str, Path], info: Dict) -> None:
+    mkdir(Path(path).parent)
+
+    with open(path, "w") as f:
+        yaml.dump(info, f)
+
```
Comment (BlueWhaleKo) [inline manta_client/util.py:45]: will fail if mkdir() has failed

File: manta_client/util.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    else:
+        try:
+            params = params.items()
+        except Exception as e:
+            print(e)
+            raise AttributeError()
+
+
+def json_value_sanitize(value):
+    return value
```
Comment (BlueWhaleKo) [inline manta_client/util.py:63]: need some documentation here :) to make it clear what is function is for

File: tests/test_config.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+def test_auto_config_default():
+    tempdir = tempfile.NamedTemporaryFile()
+    yaml_path = os.path.join(tempdir, "config_default.yaml")
+    yaml_dict = {"some": {"value": 1}, "thing": {"value": 2}}
+    mc.util.save_yaml(yaml_path, yaml_dict)
+
+    config = mc.Config()
+    tempdir.close()
+    assert dict(config) == dict(some=1, thing=2)
```
Comment (BlueWhaleKo) [inline tests/test_config.py:71]: nice 👍

File: tests/test_setting.py [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+def test_preprocess_base_url():
+    s = Settings()
+    s.update(base_url="http://host.com")
+    assert s.base_url == "http://host.com"
+    s.update(base_url="http://host.com/")
+    assert s.base_url == "http://host.com"
+    s.update(base_url="http://host.com///")
+    assert s.base_url == "http://host.com"
+    s.update(base_url="//http://host.com//")
+    assert s.base_url == "//http://host.com"
```
Comment (BlueWhaleKo) [inline tests/test_setting.py:187]: ❤️ ❤️ ❤️ ❤️

**Review by BlueWhaleKo — COMMENTED** [review]
LGTM!

**Review by BlueWhaleKo — APPROVED** [review]

File: manta_client/__init__.py (hunk 1 above)
Comment (joowon-dm-snu) [reply manta_client/__init__.py:1]: We cant :(

File: manta_client/base/config.py (hunk 2 above)
Comment (joowon-dm-snu) [reply manta_client/base/config.py:76]: sanitize process is deleted for now. 
originally used for changing key names or change tensor to real values

File: manta_client/base/config.py (hunk 3 above)
Comment (joowon-dm-snu) [reply manta_client/base/config.py:73]: added!

File: manta_client/base/setting.py (hunk 5 above)
Comment (joowon-dm-snu) [reply manta_client/base/setting.py:76]: property is object from other libs NP :)

File: manta_client/util.py (hunk 8 above)
Comment (joowon-dm-snu) [reply manta_client/util.py:45]: agreed, ill add some more error or logs later

File: manta_client/util.py (hunk 9 above)
Comment (joowon-dm-snu) [reply manta_client/util.py:63]: checked

File: manta_client/base/config.py (hunk 2 above)
Comment (joowon-dm-snu) [reply manta_client/base/config.py:76]: add TODO there

File: manta_client/base/config.py (hunk 3 above)
Comment (joowon-dm-snu) [reply manta_client/base/config.py:73]: done

File: manta_client/base/config.py (hunk 4 above)
Comment (joowon-dm-snu) [reply manta_client/base/config.py:132]: I deleted callback & its test :)

Comment (joowon-dm-snu) [reply manta_client/base/config.py:132]: I deleted callback func at config

**Pull request merged**